model_training/ml_loss.py

```diff
@@ -40,4 +40,35 @@ def max_error(F4f_pred, F4f_true):
         return 0
     else:
         return torch.max(torch.abs(F4f_pred - F4f_true))
+
+def pcgrad(task_losses, params, eps=1e-12):
+    if not params:
+        return []
+    if len(task_losses) == 0:
+        return [torch.zeros_like(p) for p in params]
```

Comment:
I would rather have assert statements than "graceful" failure. I want it to break as fast as possible if something doesn't make sense. We should never run into a situation where parms or task_losses is empty, and if we do, it is a signal that something is wrong. It should crash.